src/stub_read.s

```diff
@@ -1,47 +1,60 @@
-              .extern _Zp
+            .extern _Zp
 
-CHKIN:	      .equlab 0xffc6
-CHRIN:	      .equlab 0xffcf
-READST:	      .equlab 0xffb7
-CLRCHN:       .equlab 0xffcc
+CHKIN:	    .equlab 0xffc6
+CHRIN:	    .equlab 0xffcf
+READST:	    .equlab 0xffb7
+CLRCHN:     .equlab 0xffcc
 
-              .section code
-              .public _Stub_read
-_Stub_read:   ldx     zp:_Zp+0
-	      beq     start$	    ; stdin
-	      cpx     #3	    ; test for stdout/stderr
-	      bcc     eof$
-	      jsr     CHKIN	    ; ensure channel is prepared to read
-start$:	      lda     #0	    ; set file counter = 0
-              sta     zp:_Zp+0
-              sta     zp:_Zp+1
-              ldx     zp:_Zp+5      ; 256 or more?
-              beq     20$           ; no
-              ldy     #0
-10$:          jsr     CHRIN
-              sta     (_Zp+2),y
-              iny
-              bne     10$
-              inc     zp:_Zp+3      ; step 256
-              inc     zp:_Zp+1
-              dec     zp:_Zp+5      ; yes, decrement 256 counter
-              bne     10$
+            .section code
+            .public _Stub_read
+_Stub_read: ldx     zp:_Zp+0
+            beq     start$	    ; stdin
+            cpx     #3	    ; test for stdout/stderr
+            bcc     eof$
+            jsr     CHKIN	    ; ensure channel is prepared to read
+start$:     lda     #0	    ; set file counter = 0
+            sta     zp:_Zp+0
+            sta     zp:_Zp+1
+            lda     zp:_Zp+4
+            ora     zp:_Zp+5
+            beq     30$
+            jsr     CHRIN         ; get first byte
+            tax
+            jsr     READST
+            and     #0x02
+            bne     eof$
+            txa
+            ldy     #0
+            ldx     zp:_Zp+5      ; 256 or more?
+            beq     20$           ; no
+            bne     15$
+10$:        jsr     CHRIN
+15$:        sta     (_Zp+2),y
+            inc     zp:_Zp+0
+            jsr     READST
+            tax
+            bne     30$
+            iny
+            bne     10$
+            inc     zp:_Zp+3      ; step 256
+            inc     zp:_Zp+1
+            dec     zp:_Zp+5      ; yes, decrement 256 counter
+            bne     10$
 
-20$:          ldy     #0
-25$:          cpy     zp:_Zp+4
-              beq     30$
-              jsr     CHRIN
-              sta     (_Zp+2),y
-              inc     zp:_Zp+0
-              iny
-              bne     25$
-30$:	      jsr     READST	    ; check for error
-	      tay
-          jsr     CLRCHN
-          tya
-          tax 
-	      beq     done$
-eof$:	      lda     #-1
-	      sta     zp:_Zp+0
-	      sta     zp:_Zp+1
-done$:	      rts
+25$:        cpy     zp:_Zp+4
+            beq     30$
+            jsr     CHRIN
+20$:        sta     (_Zp+2),y
+            inc     zp:_Zp+0
+            jsr     READST
+            tax
+            bne     30$
+            iny
+            bne     25$
+30$:        jsr     CLRCHN
+            beq     done$
```

Comment:
Is this `beq` safe, can it just be an `rts`  instruction?

Reply:
I just made an update to fix that. It was working because the kernal was working with 0 as the last step of CLRCHN. But that was not contract, and could fail with any update.

I think I have found a bug that prevents returning negative numbers from _Stub_read to lib_fread, and I'll open an issue on the main project. At the moment, this always returns 0, which is at least safe.